src/Microsoft.ML.TimeSeries/RootCauseAnalyzer.cs

```diff
@@ -0,0 +1,764 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.ML.Internal.Utilities;
+
+namespace Microsoft.ML.TimeSeries
+{
+    public class RootCauseAnalyzer
+    {
+        private static double _anomalyRatioThreshold = 0.5;
+        private static double _anomalyDeltaThreshold = 0.95;
+        private static double _anomalyPreDeltaThreshold = 2;
+
+        private RootCauseLocalizationInput _src;
+        private double _beta;
+
+        public RootCauseAnalyzer(RootCauseLocalizationInput src, double beta)
+        {
+            _src = src;
+            _beta = beta;
+        }
+
+        public RootCause Analyze()
+        {
+            return AnalyzeOneLayer(_src);
+        }
+
+        //This is a function for analyze one layer for root cause, we select one dimension with values who contributes the most to the anomaly. For full result, call this function recursively
+        private RootCause AnalyzeOneLayer(RootCauseLocalizationInput src)
+        {
```

Comment:
Where is this function called recursively? Is it supposed to be called recursively by the caller? If so, why is it private?

Reply:
> Where is this function called recursively? Is it supposed to be called recursively by the caller? If so, why is it private?

There is no recursively call in the library, I will remove the word to reduce misunderstanding

Reply:
In that case, why do we need two functions Analyze and AnalyzeOneLayer? Is there a plan for more layers in the future?

---
In reply to: [422136763](https://github.com/dotnet/machinelearning/pull/4925#discussion_r422136763) [](ancestors = 422136763)

Reply:
> In that case, why do we need two functions Analyze and AnalyzeOneLayer? Is there a plan for more layers in the future?
> 
> In reply to: [422136763](https://github.com/dotnet/machinelearning/pull/4925#discussion_r422136763) [](ancestors = 422136763)

Currently, we only provide the incremental analyze, maybe in the future we will provide analyze all layers(which is actually call the AnalyzeOneLayer recursively), so we provide the private function AnalyzeOneLayer here.